### PR TITLE
Add support for concatenating var() functions in 'content' declarations

### PR DESCRIPTION
**Pull request by mikevoets**
Concatenation of `var()` functions in content declarations does not work. Currently, a variable in content declaration gets resolved correctly, however it disregards the remaining value of a content declaration. E.g.:

```
body { --var: 'Paragraph'; counter-reset: p }
p { counter-increment: p }
p:before { content: var(--var) ': ' counter(p); }
```
resolves `p:before` to `Paragraph`, and not to as you would expect; `Paragraph: 1`.

This bug is fixed in the changes in this PR.

Linked to my comment in #898.

### CODE REVIEW

**Comment by mikevoets** [comment]
Checks fail in Python 3.5 due to some unexpected syntax in `.egg` files. Any idea how these can be fixed?

File: weasyprint/css/validation/properties.py [hunk 1]
```diff
@@ -89,10 +89,11 @@ def validate_non_shorthand(base_url, name, tokens, required=False):
     if not required and name not in PROPERTIES:
         raise InvalidValues('property not supported yet')
 
-    for token in tokens:
-        var_function = check_var_function(token)
-        if var_function:
-            return ((name, var_function),)
+    if name != 'content':
```
Comment (liZe) [inline weasyprint/css/validation/properties.py:92]: Why is this change needed? I’d really like to avoid special cases based on property name…

File: weasyprint/css/computed_values.py [hunk 2]
```diff
@@ -195,6 +195,37 @@ def compute(element, pseudo_type, specified, computed, parent_style,
     """
     from .validation.properties import PROPERTIES
 
+    def resolve_var(value, name, computed):
```
Comment (liZe) [inline weasyprint/css/computed_values.py:198]: Would it be possible to put the function outside of the `compute` function? Redefining the function each time `compute` is called takes useless time.

File: weasyprint/css/computed_values.py [hunk 3]
```diff
@@ -218,37 +249,21 @@ def compute(element, pseudo_type, specified, computed, parent_style,
...
-                    value = INITIAL_VALUES[name]
-            else:
-                value = new_value
+        if value and isinstance(value, list):
+            for i, val in enumerate(value):
+                if not isinstance(val, tuple) or val[0] != 'var()':
+                    continue
+                new_value, already_computed_value = \
+                    resolve_var(val, name, computed)
+                if name == 'content':
```
Comment (liZe) [inline weasyprint/css/computed_values.py:258]: Hmmmm… There’s definitely something wrong with this. `content` shouldn’t be different from other properties. I’m not sure to understand how this PR really works, but having these special cases doesn’t look like a good solution for me :unamused:.

**Comment by liZe** [comment]
> Checks fail in Python 3.5 due to some unexpected syntax in `.egg` files. Any idea how these can be fixed?

Isort doesn’t support Python 3.5 anymore, and WeasyPrint could follow it for the next release. We can safely ignore this error for now.

File: weasyprint/css/validation/properties.py [hunk 4]
```diff
@@ -89,10 +92,11 @@ def validate_non_shorthand(base_url, name, tokens, required=False):
     if not required and name not in PROPERTIES:
         raise InvalidValues('property not supported yet')
 
-    for token in tokens:
-        var_function = check_var_function(token)
-        if var_function:
-            return ((name, var_function),)
+    if name not in MULTIVAL_PROPERTIES:
```
Comment (mikevoets) [inline weasyprint/css/validation/properties.py:95]: This is necessary because a `content` declaration is a special case since its property handler returns a list, as opposed to every other property. Currently if one of the rules in `content` is a variable function, it would only parse that variable function and return it from the function here, without processing the other rules. This is why I have added a special case condition here.

I cleaned it up by declaring the content property as a multi value property.

File: weasyprint/css/utils.py [hunk 5]
```diff
@@ -699,6 +699,11 @@ def get_content_list_token(token, base_url):
     if string is not None:
         return string
 
+    # <var>
```
Comment (mikevoets) [inline weasyprint/css/utils.py:702]: Here a variable function in a content declaration is acknowledged and returned accordingly.

File: weasyprint/css/computed_values.py [hunk 6]
```diff
@@ -218,37 +257,20 @@ def compute(element, pseudo_type, specified, computed, parent_style,
...
-                    already_computed_value = name not in INITIAL_NOT_COMPUTED
-                    value = INITIAL_VALUES[name]
-            else:
-                value = new_value
+        if value:
+            converted_to_list = False
+
+            if not isinstance(value, list):
+                converted_to_list = True
+                value = [value]
```
Comment (mikevoets) [inline weasyprint/css/computed_values.py:265]: Multi values properties (for now only content property classifies as that) return a list that possibly includes variable functions among their declaration values. Therefore, doing `isinstance(value, tuple) and value[0] == 'var()'` would never compute a variable for that situation.

I fixed this by looping over all the values and compute the variables when there are variable functions. I temporarily convert non-list declaration values to a list to do this elegantly.

File: weasyprint/css/computed_values.py [hunk 7]
```diff
@@ -176,6 +176,46 @@ def decorator(function):
...
+                'for property `%s`.', computed_value,
+                variable_name.replace('_', '-'), name.replace('_', '-'))
+            if name in INHERITED and parent_style:
+                already_computed_value = True
+                value = parent_style[name]
+            else:
+                already_computed_value = name not in INITIAL_NOT_COMPUTED
+                value = INITIAL_VALUES[name]
+        elif isinstance(new_value, list):
+            value, = new_value
```
Comment (mikevoets) [inline weasyprint/css/computed_values.py:213]: The `content` property handler always returns a list regardless of input. In this case, the handler will only ever get a computed variable declaration, and thus I unpack the computed value again after getting it back from the handler.

**Comment by mikevoets** [comment]
@liZe I've replied to your comments in my newest commit. I have also cleaned up a whole bit. Thanks for your review and feedback!

**Comment by mikevoets** [comment]
@liZe You're probably busy, but did you have time to look at this?

**Comment by liZe** [comment]
> @liZe You're probably busy, but did you have time to look at this?

I’m busy, as I’m… on holidays :smile:. Thanks a lot for your changes, I’ll have the time to review them this week or next week.

**Comment by mikevoets** [comment]
Hope you had a good holiday @liZe. If there's anything I can do to improve the PR, let me know :-)

**Comment by liZe** [comment]
> Hope you had a good holiday @liZe. If there's anything I can do to improve the PR, let me know :-)

I’m late, I’m really sorry… It’s currently complicated for me to find the time for a good review. I’ll do my best…

**Comment by mikevoets** [comment]
> > Hope you had a good holiday @liZe. If there's anything I can do to improve the PR, let me know :-)
> 
> I’m late, I’m really sorry… It’s currently complicated for me to find the time for a good review. I’ll do my best…

All good! Hope you'll be able to look at it soon. Weasyprint has been extremely useful in my projects and I am happy to continuously contribute to improve it.

**Comment by mikevoets** [comment]
Hi @liZe, just checking in to see if you had any chance to reviewing this PR any time soon? I'm depending on this fix currently, and I need to use a fork of this repository until the PR has been merged. Hope I can revert to upstream soon!

**Comment by liZe** [comment]
> Hi @liZe, just checking in to see if you had any chance to review this PR any time soon? I'm depending on this fix currently, and I need to use a fork of this repository until the PR has been merged. Hope I can revert to upstream soon!

Thanks again for your patience.

I’ve found the time to read your pull request, and I have mixed feelings about it. Your code definitely goes in the right direction, I will merge it, and I’ll add a couple of minor changes (for example, to add all the properties returning iterables).

The main problem with this PR is not your code 😄. The main problem is that this PR shows how much the current implementation is wrong. I already knew that a lot of CSS values are stored in random Python structures, and passed from function to function where they’re randomly detected and randomly computed. It’s been bad for a long time, but it now reaches dangerous limits, and I don’t want to go beyond.

Let’s say this PR is the last one before a large rewrite. If we want to have `var`, `attr` and `calc` (hello #357!) working correctly, we have to clean this quickly before it turns into a monster (I’m looking at you, `split_first_line`).

**Comment by mikevoets** [comment]
Thanks so much. If there's anything I can do to help you cleaning up, let me know.

**Comment by liZe** [comment]
Thanks again, it’s finally merged.

> I will merge it, and I’ll add a couple of minor changes (for example, to add all the properties returning iterables).

Most of the properties returning iterables use the `comma_separated_list` decorator and return a tuple. Having a single way to handle `var` for `content` and these properties would require a lot of work.

We’ll rewrite this sooner or later, to get a clean support of `var`, `calc` and `attr`.

**Comment by mikevoets** [comment]
@liZe thanks!
